### ISSUE TITLE
A psi-rule may be executed twice

### ISSUE THREAD

**Issue by leungmanhin** [opencog/opencog#2608]
This is actually known for a while, I'm creating this issue to keep track of it.

Observations:
- it happens only when we're running the whole HR stack, couldn't reproduce it in a clean cogserver
- it happens only when it's evaluating some simple / lightweight psi-rule -- a psi-rule that doesn't take much computing power to be evaluated and executed

Below is a way to reproduce it, with the whole HR stack running:
```scheme
; Disable all chat-related psi-rules just to make the result clear
(disable-all-demos)

; Create a simple psi-rule
(psi-rule
    (list (SequentialAnd
        (DefinedPredicate "is-input-utterance?")
    ))
    (True (ExecutionOutput (GroundedSchema "scm: say") (List (Word "hi"))))
    (True)
    (stv .9 .9)
    sociality
)
```

then chat with it, on my computer it keeps saying "hi" twice for each input most of the time

and from the log, that rule actually got evaluated (and executed) twice in two different iterations of the psi-loop

**Comment by leungmanhin** [comment]
Some updates to this issue:
The `(chat)` function in `chatbot-psi` is indeed being called twice, for every input, and this probably is the cause of this problem. It doesn't cause the chatbot to reply twice all the time because it depends on the execution sequences of the codes, which varies all the time.

Here is one failing case I saw that caused it to reply twice:
1) chatbot-psi was called for INPUT
2) chatbot-psi was called for INPUT again
3) States were updated so that `(DefinedPredicate "is-input-utterance?")` was satisfied
4) OpenPsi finished the evaluation the psi-rule, and a reply was generated
5) States were updated again so that `(DefinedPredicate "is-input-utterance?")` was satisfied
6) OpenPsi finished the evaluation the rule, and another reply was generated

On the other hand, if states were updated before the evaluation (i.e. both step 3 and step 5 being executed before step 4), then it would only reply once for the input.
